### ISSUE TITLE
Twisted.python.win32.cmdLineQuote doesn't correctly quote pipes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3933]
|<img alt="dberger@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| dberger@... reported|
|-|-|
|Trac ID|trac#3933|
|Type|defect|
|Created|2009-07-27 20:48:24Z|
|Branch|https://github.com/twisted/twisted/tree/win32-pipe-quoting-3933-2|

we're using twisted via buildbot, and running builds on win32 using Incredibuild, which requires the buildconfiguration to be specified as config|platform (i.e. "debug|win32"), even if only one platform exists in the project.

the cmdLineQuote method on Twisted.python.win32 doesn't correctly quote the pipe character.

adding the pipe to the set of characters tested for fixes the problem.

<details><summary>Searchable metadata</summary>

```
trac-id__3933 3933
type__defect defect
reporter__dberger____ dberger@...
priority__high high
milestone__ 
branch__branches_win32_pipe_quoting_3933_2 branches/win32-pipe-quoting-3933-2
branch_author__jonathanj jonathanj
status__closed closed
resolution__invalid invalid
component__core core
keywords__win32 win32
time__1248727704000000 1248727704000000
changetime__1264628277000000 1264628277000000
version__None None
owner__ 
cc__dberger@... cc__ivank cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Thanks for noticing this.  You wouldn't happen to have a patch with a test, would you? :)

**Comment by twisted-trac** [comment]
|<img alt="dberger@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|dberger@... commented|
|-|-|

Sorry, no - we fixed it in our local p4 repository, but no test, and the "patch" is trivial, add "or ('|' in s)" to the test in cmdLineQuote

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Replying to [dberger@...](#note_2):
> Sorry, no - we fixed it in our local p4 repository, but no test, and the "patch" is trivial, add "or ('|' in s)" to the test in cmdLineQuote

I know; I assumed that if you had one it would already have been attached.  It was supposed to be a not-so-subtle hint.  I'll be even more unsubtle: it would be very helpful if you could submit such a patch for review, according to [the steps outlined in our development process](https://twistedmatrix.com/trac/wiki/TwistedDevelopment).  If you are interested in contributing to Twisted it would also be a great starting point.

If not, that's fine, just noticing the bug was helpful.  It just might take us a while to get to since we are perpetually shorthanded.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_4"></a>|@jonathanj commented|
|-|-|

(In [27215]) Branching to 'win32-pipe-quoting-3933'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I don't believe that the "easy" keyword can apply to any Windows ticket.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_7"></a>|@jonathanj commented|
|-|-|

(In [27222]) Fix test_quoteArguments breaking on arguments containing whitespace, refs [#3933](https://github.com/twisted/twisted/issues/3933).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|ivank commented|
|-|-|

Here is some background information:

```
The ampersand (&), pipe (|), and parentheses ( ) are special characters that
must be preceded by the escape character (^) or quotation marks when you pass
them as arguments."
```

http://technet.microsoft.com/en-us/library/cc737438%28WS.10%29.aspx

http://www.microsoft.com/resources/documentation/windows/xp/all/proddocs/en-us/ntcmds_shelloverview.mspx



I like this patch. The structure of the tests, and the cleanups make sense to me. Here are some comments:

1. Every argument should be quoted, or all `^` symbols must be doubled. This is because an application might pass the escape character `^` in an argument.
```
C:\>python -c "import sys; print sys.argv" hello^world
['-c', 'helloworld']

C:\>python -c "import sys; print sys.argv" "hello^world"
['-c', 'hello^world']

C:\>python -c "import sys; print sys.argv" hello^^world
['-c', 'hello^world']
```

This could breaks backwards compatibility with applications that are using `^` as a workaround for `cmdLineQuote` not doing its job. I think it would be best to fix it anyway. This incompatibility should be documented somewhere, but I don't know where.

2. The characters `&()` should be quoted/escaped as well. It doesn't look like quoting/escaping `(` or `)` is necessary on XP or Vista, but it can't do any harm.

3. Instead of 
```
try:
    set
except NameError:
    from sets import Set as set
```
most Twisted code uses `from twisted.python.compat import set`. I think this is still the right thing to do.

4. The choice of characters that `_cmdLineQuoteChars` uses should be documented, maybe with a link to a page on microsoft.com

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_10"></a>|@jonathanj commented|
|-|-|

(In [27245]) Enclose every argument in double quotes, removing the need for a list of
special characters.

refs [#3933](https://github.com/twisted/twisted/issues/3933)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj set owner to ivan|
|-|-|

The code is a lot simpler now that everything is just being double-quote quoted. I also add a link to a helpful article on the topic.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj set owner to ivank|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

This change is definitely pretty cool.  It's much simpler than the previous implementation.  If it also has the advantage of working properly, then it will be a complete win. :)

So... about that working properly...  I think it would be best if we actually exercised this piece of OS integration.  Can you add a test which really launches a child process and verifies that all bytes (except NUL, I guess?) arrive in the child properly?  This should show us both that the previous implementation _was_ broken for the "|" case (as well, perhaps, as for others) and that the new implementation accounts for that case.  It also has the benefit of showing us if Windows ever changes the rules on us (not likely, but not impossible either).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jonathanj|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_15"></a>|@jonathanj commented|
|-|-|

(In [27247]) Integration testing for argument quoting on win32.

refs [#3933](https://github.com/twisted/twisted/issues/3933)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_17"></a>|@jonathanj commented|
|-|-|

(In [27287]) Non-NUL, really.

refs [#3933](https://github.com/twisted/twisted/issues/3933)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_18"></a>|@jonathanj commented|
|-|-|

(In [27288]) Use twisted.internet.utils.getProcessOutput.

refs [#3933](https://github.com/twisted/twisted/issues/3933)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. `QuotingTests.test_integration` should be runnable on all platforms with all reactors.  Check out `twisted.internet.test.test_process`.  The Windows-specific logic in it is unfortunate and we should perhaps do something about it eventually.  What _is_ the guarantee about the stream type `spawnProcess` gives you?  Apparently it's text.  Needs doc'in.  Anyway, for now we should still run this test everywhere.  Conditionalize the Windows stuff to just run on Windows, but the rest of the test should run everywhere.
  1. It's so sad that these are the first tests for `twisted.python.win32`.  Sigh.  Oh well.  Please add a copyright statement and module docstring to the new test module.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_21"></a>|@jonathanj commented|
|-|-|

(In [27293]) Perform command-line argument quoting integration tests on all platforms.

Also added a copyright statement and module docstring for new
"twisted.python.test.test_win32" module.

refs [#3933](https://github.com/twisted/twisted/issues/3933)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. `test_win32.py` didn't exist prior to 2009. ;)  The copyright date in it should just be 2009.
  1. The docstring for `test_commandLineArgumentQuoting` is overly specific.  It talks about quoting, but Windows is the only platform where quoting is actually done.  It should just talk about the bebavior, not the implementation. (Similarly, the name is overly specific, I suppose.)
  1. If the assertion in `test_commandLineArgumentQuoting` fails, then I think the test will hang until it times out.  This could probably be improved in one of several ways.  The assertion could happen in the main test method, after the runReactor call.  Or the reactor.stop could happen in an addBoth callback, or something like that.

Basically looks good now though, hopefully this will be the last round of feedback. :)  Thanks for your work on this.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_24"></a>|@jonathanj commented|
|-|-|

(In [27408]) Fix copyright date. Improve docstring and behavour for integration tests. refs [#3933](https://github.com/twisted/twisted/issues/3933)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @jonathanj|
|-|-|

There's a conflict in `test_process.py`, can you merge forward and resolve?

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_27"></a>|@jonathanj commented|
|-|-|

(In [27419]) Branching to 'win32-pipe-quoting-3933-2'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_28"></a>|@jonathanj set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_26):
> There's a conflict in `test_process.py`, can you merge forward and resolve?

Done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

Okay, so after some more careful investigation, including tracking down the source of this report to http://buildbot.net/trac/ticket/377 and far more time than I really cared to spend reading documentation about Windows brain damage, my conclusion is that this isn't a valid bug.

The API Twisted wraps, `CreateProcess`, does not require | to be quoted.  Quoting does no harm, but neither does it do any good.  Instead, it is certain other programs which have quoting rules of which Twisted is presently unaware.  For example, cmd.exe does require | to be quoted.  `spawnProcess` is the wrong place to fix cmd.exe quoting issues, though.  Instead, if a nice, error-resistant cmd.exe-invocation API is desired, it should be layered on top of `spawnProcess` and the appropriate quoting rules added there.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_30"></a>|ivank commented|
|-|-|

Just linking to exarkun's discovery, http://www.autohotkey.net/~deleyd/parameters/parameters.htm

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_31"></a>|ivank commented|
|-|-|

jonathanj's new test `test_commandLineArgumentQuoting` does indeed pass on trunk's `quoteArguments`/`cmdLineQuote`. Sorry for the earlier incorrect review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

This was a duplicate of [#10739](https://github.com/twisted/twisted/issues/10739).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_33"></a>|@khorn commented|
|-|-|

see also: [#3876](https://github.com/twisted/twisted/issues/3876) regarding empty commnad line args not being quoted properly on Win32.

the issue was mentioned in the duplicate ticket ([#10739](https://github.com/twisted/twisted/issues/10739)), so I thought I'd mention it here as well for those interested in this sort of thing

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
